File: Source/AlphaTab/Rendering/Glyphs/TabNoteChordGlyph.cs
/*
 * This file is part of alphaTab.
 * Copyright © 2017, Daniel Kuschny and Contributors, All rights reserved.
 * 
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3.0 of the License, or at your option any later version.
 * 
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library.
 */

using AlphaTab.Collections;
using AlphaTab.Model;
using AlphaTab.Platform;
using AlphaTab.Platform.Model;
using AlphaTab.Rendering.Utils;

namespace AlphaTab.Rendering.Glyphs
{
    public class TabNoteChordGlyph : Glyph
    {
        private readonly FastList<NoteNumberGlyph> _notes;
        private readonly bool _isGrace;

        public Beat Beat { get; set; }
        public BeamingHelper BeamingHelper { get; set; }
        public Note MinStringNote { get; set; }
        public FastDictionary<string, Glyph> BeatEffects { get; set; }
        public FastDictionary<int, NoteNumberGlyph> NotesPerString { get; set; }

        public TabNoteChordGlyph(float x, float y, bool isGrace)
            : base(x, y)
        {
            _isGrace = isGrace;
            _notes = new FastList<NoteNumberGlyph>();
            BeatEffects = new FastDictionary<string, Glyph>();
            NotesPerString = new FastDictionary<int, NoteNumberGlyph>();
        }

        public float GetNoteX(Note note, bool onEnd = true)
        {
            if (NotesPerString.ContainsKey(note.String))
            {
                var n = NotesPerString[note.String];
                var pos = X + n.X;
                if (onEnd)
                {
                    pos += n.Width;
                }
                return pos;
            }
            return 0;
        }

        public float GetNoteY(Note note)
        {
            if (NotesPerString.ContainsKey(note.String))
            {
                return Y + NotesPerString[note.String].Y;
            }
            return 0;
        }

        public override void DoLayout()
        {
            var w = 0f;
            for (int i = 0, j = _notes.Count; i < j; i++)
            {
                var g = _notes[i];
                g.Renderer = Renderer;
                g.DoLayout();
                if (g.Width > w)
                {
                    w = g.Width;
                }
            }

            var tabHeight = Renderer.Resources.TablatureFont.Size;
            var effectY = GetNoteY(MinStringNote) + tabHeight / 2;
            // TODO: take care of actual glyph height
            var effectSpacing = 7 * Scale;
            foreach (var beatEffectKey in BeatEffects)
            {
                var g = BeatEffects[beatEffectKey];
                g.Y += effectY;
                g.X += Width / 2;
                g.Renderer = Renderer;
                effectY += effectSpacing;
                g.DoLayout();
            }

            Width = w;
        }

        public void AddNoteGlyph(NoteNumberGlyph noteGlyph, Note note)
        {
            _notes.Add(noteGlyph);
            NotesPerString[note.String] = noteGlyph;
            if (MinStringNote == null || note.String < MinStringNote.String) MinStringNote = note;
        }

        public override void Paint(float cx, float cy, ICanvas canvas)
        {
            cx += X;
            cy += Y;

            var res = Renderer.Resources;
            var oldBaseLine = canvas.TextBaseline;
            canvas.TextBaseline = TextBaseline.Middle;
            canvas.Font = _isGrace ? res.GraceFont : res.TablatureFont;
            var notes = _notes;
            var w = Width;
            foreach (var g in notes)
            {
                g.Renderer = Renderer;
                g.Width = w;
                g.Paint(cx, cy, canvas);
            }
            canvas.TextBaseline = oldBaseLine;

            foreach (var beatEffectKey in BeatEffects)
            {
                var g = BeatEffects[beatEffectKey];
                g.Paint(cx, cy, canvas);
            }
        }

        public void UpdateBeamingHelper(float cx)
        {
            if (BeamingHelper != null && BeamingHelper.IsPositionFrom(TabBarRenderer.StaffId, Beat))
            {
                BeamingHelper.RegisterBeatLineX(TabBarRenderer.StaffId, Beat, cx + X + Width, cx + X);
            }
        }
    }
}

File: src/rendering/glyphs/TabNoteChordGlyph.ts
import { type Beat, BeatSubElement } from '@src/model/Beat';
import type { Note } from '@src/model/Note';
import { type ICanvas, TextBaseline } from '@src/platform/ICanvas';
import { Glyph } from '@src/rendering/glyphs/Glyph';
import type { NoteNumberGlyph } from '@src/rendering/glyphs/NoteNumberGlyph';
import type { BeamingHelper } from '@src/rendering/utils/BeamingHelper';
import type { RenderingResources } from '@src/RenderingResources';
import { NoteXPosition, NoteYPosition } from '@src/rendering/BarRendererBase';
import type { BeatBounds } from '@src/rendering/utils/BeatBounds';
import { DeadSlappedBeatGlyph } from '@src/rendering/glyphs/DeadSlappedBeatGlyph';
import { ElementStyleHelper } from '@src/rendering/utils/ElementStyleHelper';

export class TabNoteChordGlyph extends Glyph {
    private _notes: NoteNumberGlyph[] = [];
    private _deadSlapped: DeadSlappedBeatGlyph | null = null;
    private _isGrace: boolean;

    public beat!: Beat;
    public beamingHelper!: BeamingHelper;
    public minStringNote: Note | null = null;
    public beatEffects: Map<string, Glyph> = new Map();
    public notesPerString: Map<number, NoteNumberGlyph> = new Map();
    public noteStringWidth: number = 0;

    public constructor(x: number, y: number, isGrace: boolean) {
        super(x, y);
        this._isGrace = isGrace;
    }

    public buildBoundingsLookup(beatBounds: BeatBounds, cx: number, cy: number) {
        for (const note of this._notes) {
            note.buildBoundingsLookup(beatBounds, cx + this.x, cy + this.y);
        }
    }

    public getNoteX(note: Note, requestedPosition: NoteXPosition): number {
        if (this.notesPerString.has(note.string)) {
            const n = this.notesPerString.get(note.string)!;

            let pos = this.x + n.x;
            switch (requestedPosition) {
                case NoteXPosition.Left:
                    break;
                case NoteXPosition.Center:
                    pos += n.noteStringWidth / 2;
                    break;
                case NoteXPosition.Right:
                    pos += n.width;
                    break;
            }
            return pos;
        }
        return 0;
    }

    public getNoteY(note: Note, requestedPosition: NoteYPosition): number {
        if (this.notesPerString.has(note.string)) {
            const n = this.notesPerString.get(note.string)!;
            let pos = this.y + n.y;

            switch (requestedPosition) {
                case NoteYPosition.Top:
                case NoteYPosition.TopWithStem:
                    pos -= n.height / 2 + 2;
                    break;
                case NoteYPosition.Center:
                    break;
                case NoteYPosition.Bottom:
                case NoteYPosition.BottomWithStem:
                    pos += n.height / 2;
                    break;
            }

            return pos;
        }
        return 0;
    }

    public override doLayout(): void {
        let w: number = 0;

        if (this.beat.deadSlapped) {
            this._deadSlapped = new DeadSlappedBeatGlyph();
            this._deadSlapped.renderer = this.renderer;
            this._deadSlapped.doLayout();
            w = this._deadSlapped.width;
            this.noteStringWidth = w;
        } else {
            let noteStringWidth: number = 0;
            for (let i: number = 0, j: number = this._notes.length; i < j; i++) {
                const g: NoteNumberGlyph = this._notes[i];
                g.renderer = this.renderer;
                g.doLayout();
                if (g.width > w) {
                    w = g.width;
                }
                if (g.noteStringWidth > noteStringWidth) {
                    noteStringWidth = g.noteStringWidth;
                }
            }
            this.noteStringWidth = noteStringWidth;
            const tabHeight: number = this.renderer.resources.tablatureFont.size;
            let effectY: number = this.getNoteY(this.minStringNote!, NoteYPosition.Center) + tabHeight / 2;
            // TODO: take care of actual glyph height
            const effectSpacing: number = 7;
            for (const g of this.beatEffects.values()) {
                g.y += effectY;
                g.x += this.width / 2;
                g.renderer = this.renderer;
                effectY += effectSpacing;
                g.doLayout();
            }
        }

        this.width = w;
    }

    public addNoteGlyph(noteGlyph: NoteNumberGlyph, note: Note): void {
        this._notes.push(noteGlyph);
        this.notesPerString.set(note.string, noteGlyph);
        if (!this.minStringNote || note.string < this.minStringNote.string) {
            this.minStringNote = note;
        }
    }

    public override paint(cx: number, cy: number, canvas: ICanvas): void {
        cx += this.x;
        cy += this.y;

        if (this.beat.deadSlapped) {
            this._deadSlapped?.paint(cx, cy, canvas);
        } else {
            const res: RenderingResources = this.renderer.resources;
            const oldBaseLine: TextBaseline = canvas.textBaseline;
            canvas.textBaseline = TextBaseline.Middle;
            canvas.font = this._isGrace ? res.graceFont : res.tablatureFont;

            const notes: NoteNumberGlyph[] = this._notes;
            const w: number = this.width;
            for (const g of notes) {
                g.renderer = this.renderer;
                g.width = w;
                g.paint(cx, cy, canvas);
            }
            canvas.textBaseline = oldBaseLine;

            using _ = ElementStyleHelper.beat(canvas, BeatSubElement.GuitarTabEffects, this.beat);
            for (const g of this.beatEffects.values()) {
                g.paint(cx, cy, canvas);
            }
        }
    }

    public updateBeamingHelper(cx: number): void {
        if (this.beamingHelper && this.beamingHelper.isPositionFrom('tab', this.beat)) {
            this.beamingHelper.registerBeatLineX(
                'tab',
                this.beat,
                cx + this.x + this.width / 2,
                cx + this.x + this.width / 2
            );
        }
    }
}
